plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php

```diff
@@ -570,4 +570,217 @@ private function validate_data( $value, string $type ) {
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared -- $query should already be prepared.
+		$results = $wpdb->get_results( $query, ARRAY_A );
+
+		return $this->verify_data( $results );
+	}
+
+	/**
+	 * Generate query to fetch data from both source and destination tables. Use the results in `verify_data` to verify if data was migrated properly.
+	 *
+	 * @param array $source_ids Array of IDs in source table.
+	 *
+	 * @return string SELECT statement.
+	 */
+	protected function build_verification_query( $source_ids ) {
+		$source_table                  = $this->schema_config['source']['entity']['table_name'];
+		$meta_table                    = $this->schema_config['source']['meta']['table_name'];
+		$destination_table             = $this->schema_config['destination']['table_name'];
+		$meta_entity_id_column         = $this->schema_config['source']['meta']['entity_id_column'];
+		$meta_key_column               = $this->schema_config['source']['meta']['meta_key_column'];
+		$meta_value_column             = $this->schema_config['source']['meta']['meta_value_column'];
+		$destination_source_rel_column = $this->schema_config['destination']['source_rel_column'];
+		$source_destination_rel_column = $this->schema_config['source']['entity']['destination_rel_column'];
+		$source_meta_rel_column        = $this->schema_config['source']['entity']['meta_rel_column'];
+
+		$source_destination_join_clause = "$destination_table ON $destination_table.$destination_source_rel_column = $source_table.$source_destination_rel_column";
+
+		$meta_select_clauses        = array();
+		$meta_join_clauses          = array();
+		$source_select_clauses      = array();
+		$destination_select_clauses = array();
+
+		foreach ( $this->core_column_mapping as $column_name => $schema ) {
+			$source_select_column         = isset( $schema['select_clause'] ) ? $schema['select_clause'] : "$source_table.$column_name";
+			$source_select_clauses[]      = "$source_select_column as {$source_table}_{$column_name}";
+			$destination_select_clauses[] = "$destination_table.{$schema['destination']} as {$destination_table}_{$schema['destination']}";
+		}
+
+		foreach ( $this->meta_column_mapping as $meta_key => $schema ) {
+			$meta_table_alias             = "meta_source_{$schema['destination']}";
+			$meta_select_clauses[]        = "$meta_table_alias.$meta_value_column AS $meta_table_alias";
+			$meta_join_clauses[]          = "
+$meta_table $meta_table_alias ON
+	$meta_table_alias.$meta_entity_id_column = $source_table.$source_meta_rel_column AND
+	$meta_table_alias.$meta_key_column = '$meta_key'
+";
+			$destination_select_clauses[] = "$destination_table.{$schema['destination']} as {$destination_table}_{$schema['destination']}";
+		}
+
+		$select_clause = implode( ', ', array_merge( $source_select_clauses, $meta_select_clauses, $destination_select_clauses ) );
+
+		$meta_join_clause = implode( ' LEFT JOIN ', $meta_join_clauses );
```

Comment:
Just double checking here: This can potentially create a query with quite a few `JOIN`s, e.g. the address verification joins postmeta 11 times, the operational verification 15 times. Do you think it might be potentially problematic on sites with huge postmeta (e.g. 1M+ rows)? That said, the speed of this is impressive: on my site with 800k postmeta records (around 0.1s) and EXPLAIN looks cool and searches very few rows, so it seems to be okay.

Reply:
Yeah, this makes for a rather large joint. An alternative approach exists, in which we would fetch metadata, core data, and migrated data in three different queries. We use this approach in the migration logic, if there are issues here, we can change this to use that other approach. But let's keep this as it is for now, that way we will also test both of these approaches and see if one is better than the other. wdyt?

Otherwise it should be fast enough since there are where conditions on both ids as well as meta_keys, so MySQL use either of indexes.

Reply:
Yeah, let's keep it and see what we get back from testing. 👍